Cast Details Screen

skaik-mo opened the pull request:
Key Updates:

    Added CastDetailsScreen, CastDetailsContent, and CastDetailsViewModel.

    Introduced CastDetailsUiState and CastDetailsInteractionListener.

    Reused BaseViewModel for shared logic.

    Enhanced MoviesListSection and GallerySection to support empty states.

    Refactored PosterItemVertically and MainDetails layout.

    Registered CastDetailsViewModel in viewModelModule.

Video Preview included below.

https://github.com/user-attachments/assets/35fc815c-dc46-4f16-ac0c-d2c9313eb18f

skaik-mo committed:
add the cast details screen

Key changes include:
- Created `CastDetailsScreen.kt` for displaying actor details.
- Implemented `CastDetailsContent` with scroll-aware animated header and sections for movies, gallery, and biography.
- Added `CastDetailsViewModel.kt` to manage the UI state and handle user interactions.
- Created `CastDetailsUiState.kt` to represent the data displayed on the screen.
- Created `CastDetailsInteractionListener.kt` for handling user actions.
- Added `BaseViewModel.kt` for common ViewModel functionalities.
- Updated `MoviesListSection` to handle empty movie lists and pass movie IDs on click.
- Updated `GallerySection` to handle empty image lists and use String URLs for images.
- Refactored `PosterItemVertically` to move the click listener to the image.
- Modified `MainDetails.kt` to use `HorizontalDivider` and adjust padding.
- Added `CastDetailsViewModel` to the Koin `viewModelModule`.

Mostafa-khaled21 commented on presentation/details/src/main/java/com/giraffe/details/baseViewModel/BaseViewModel.kt:
```diff
@@ -0,0 +1,49 @@
+
+open class BaseViewModel<STATE>(initialState: STATE) : ViewModel() {
+
+    protected val _state: MutableStateFlow<STATE> by lazy { MutableStateFlow(initialState) }
```
You don't need to use by lazy here

Mostafa-khaled21 commented on presentation/details/src/main/java/com/giraffe/details/components/GallerySection.kt:
```diff
@@ -25,9 +24,12 @@ import com.giraffe.details.utils.imageSourceToPainter
 fun GallerySection(
     modifier: Modifier = Modifier,
-    images: List<Pair<Int?, String?>>,
+    images: List<Pair<String?, String?>>,
```
Using Pair makes it unclear what each value represents (first and second don’t convey meaning).Replacing it with a data class.

skaik-mo committed:
Refactor: Make _state in BaseViewModel private

The `_state` MutableStateFlow in `BaseViewModel` is now private to encapsulate its modification within the ViewModel. The `lazy` initialization has also been removed as it's not strictly necessary for this use case.

skaik-mo committed:
Update GallerySection to accept separate lists for images and content descriptions

The key changes are:

-   `GallerySection` now accepts two separate lists: `images` (List<String?>) and `imageContentDescriptions` (List<String?>) instead of a single list of Pairs.
-   The `CastDetailsUiState` is updated to reflect this change, providing `actorGalleryImages` and `actorGalleryImagesDescriptions` as separate lists.
-   The `CastDetailsScreen` is updated to pass the new `imageContentDescriptions` to the `GallerySection`.

poula22 commented on presentation/details/src/main/java/com/giraffe/details/screens/castDetails/CastDetailsScreen.kt:
```diff
@@ -0,0 +1,193 @@
+@OptIn(ExperimentalSharedTransitionApi::class)
+@Composable
+fun CastDetailsContent(
+    state: CastDetailsUiState,
+    modifier: Modifier = Modifier,
+    onYoutubeClick: () -> Unit,
+    onFacebookClick: () -> Unit,
+    onInstagramClick: () -> Unit,
+    onMoviePosterClick: (movieId: Int) -> Unit,
+    onShowMoreMoviesClick: () -> Unit,
+    onShowMoreActorGalleryClick: () -> Unit,
```
Modifier should be first optional parameter:
```suggestion
    state: CastDetailsUiState,
    onYoutubeClick: () -> Unit,
    onFacebookClick: () -> Unit,
    onInstagramClick: () -> Unit,
    onMoviePosterClick: (movieId: Int) -> Unit,
    onShowMoreMoviesClick: () -> Unit,
    onShowMoreActorGalleryClick: () -> Unit,
    modifier: Modifier = Modifier,
```

poula22 reviewed (approved):
Great worlk👏
